File: cpp/data_structures/li_chao.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
const ll LICH_INF = (1LL << 62);
struct LiChao {
  struct Line {
    ll m, b;
    ll get(ll x) const {
      return m * x + b;
    }
  };
  struct Node {
    Line ln;
    Node *l = 0, *r = 0;
    Node(Line ln) : ln(ln) {}
  };
  ll lo, hi;
  Node *root = 0;
  LiChao(ll l, ll r) : lo(l), hi(r) {}
  void add(Line nw) {
    add(root, lo, hi, nw);
  }
  void add(Node *&t, ll l, ll r, Line nw) {
    if (!t) {
      t = new Node(nw);
      return;
    }
    ll m = (l + r) / 2;
    bool lef = nw.get(l) < t->ln.get(l), mid = nw.get(m) < t->ln.get(m);
    if (mid)
      swap(nw, t->ln);
    if (l == r)
      return;
    if (lef != mid)
      add(t->l, l, m, nw);
    else
      add(t->r, m + 1, r, nw);
  }
  ll query(ll x) {
    return query(root, lo, hi, x);
  }
  ll query(Node *t, ll l, ll r, ll x) {
    if (!t)
      return LICH_INF;
    ll ans = t->ln.get(x);
    if (l == r)
      return ans;
    ll m = (l + r) / 2;
    return min(ans, x <= m ? query(t->l, l, m, x) : query(t->r, m + 1, r, x));
  }
};

File: python/data_structures/li_chao.py
INF = 10**30


class Line:
    def __init__(self, m=0, b=INF):
        self.m = m
        self.b = b

    def __call__(self, x):
        return self.m * x + self.b


class LiChao:
    def __init__(self, xs):
        self.xs = sorted(set(xs))
        self.n = len(self.xs)
        self.st = [Line() for _ in range(4 * max(1, self.n))]

    def add_line(self, line, p=1, l=0, r=None):
        if r is None:
            r = self.n
        m = (l + r) // 2
        x_l, x_m, x_r = self.xs[l], self.xs[m], self.xs[r - 1]
        cur = self.st[p]
        if line(x_m) < cur(x_m):
            self.st[p], line = line, cur
        if r - l == 1:
            return
        if line(x_l) < self.st[p](x_l):
            self.add_line(line, p << 1, l, m)
        elif line(x_r) < self.st[p](x_r):
            self.add_line(line, p << 1 | 1, m, r)

    def query(self, x, p=1, l=0, r=None):
        if r is None:
            r = self.n
        res = self.st[p](x)
        if r - l == 1:
            return res
        m = (l + r) // 2
        if x < self.xs[m]:
            return min(res, self.query(x, p << 1, l, m))
        return min(res, self.query(x, p << 1 | 1, m, r))
